upf/upf_api.c

```diff
@@ -1061,8 +1061,49 @@ vl_api_upf_tdf_ul_table_add_t_handler (vl_api_upf_tdf_ul_table_add_t * mp)
   REPLY_MACRO (VL_API_UPF_TDF_UL_TABLE_ADD_REPLY);
 }
 
-#include <upf/upf.api.c>
+/* API message handler */
+static void
+vl_api_upf_nat_pool_add_t_handler (vl_api_upf_nat_pool_add_t * mp)
+{
+  int rv = 0;
+  upf_main_t *sm = &upf_main;
+  vl_api_upf_nat_pool_add_reply_t *rmp = NULL;
+
+  u8 *name_vec = 0;
+  u8 *nwi_vec = 0;
+  u32 name_vec_len = 0;
+  ip4_address_t start, end;
+
+  mp->min_port = ntohs (mp->min_port);
+  mp->max_port = ntohs (mp->max_port);
+  mp->block_size = ntohl (mp->block_size);
+  do
+    {
+      if (mp->nwi_len > 64 || mp->name_len > 64)
+	{
+	  rv = VNET_API_ERROR_INVALID_VALUE;
+	  break;
+	}
+      vec_validate (nwi_vec, mp->nwi_len);
+      memcpy (nwi_vec, mp->nwi, mp->nwi_len);
+
+      vec_validate (name_vec, mp->name_len);
+      memcpy (name_vec, mp->name, mp->name_len);
+
+      ip4_address_decode (mp->start, &start);
+      ip4_address_decode (mp->end, &end);
+
+      rv =
+	vnet_upf_nat_pool_add_del (nwi_vec, start, end, name_vec,
+				   mp->block_size, mp->min_port, mp->max_port,
+				   mp->is_add);
```

Comment:
Note that `vnet_upf_nat_pool_add_del()` doesn't take ownership of vectors you pass to it, instead cloning their contents:
```
      nat_pool->name = vec_dup (name);
      nat_pool->network_instance = vec_dup (nwi_name);
```
which means we've got leak here, and must free `nwi_vec` and `name_vec` afterwards to fix it